website/blog/history.md

```diff
@@ -0,0 +1,466 @@
+[mTLS][mtls] to ensure that components only communicate with the components to
+which they have references.
+
+## May 30, 203: [Introduce Main Method](https://github.com/ServiceWeaver/weaver/commit/570d1ee2dfda2f245e34d9b97ffbd921ba4cd7f2)
+
+We started to think that the signature of `weaver.Run` was a bit too
+complicated. It was difficult to understand unless you were very familiar with
+Go generics:
+
+```go
+type PointerToMain[T any] interface {
+    *T
+    InstanceOf[Main]
+}
+func Run[T any, P PointerToMain[T]](ctx context.Context, app func(context.Context, *T) error) error
+```
+
+So, we added a `Main` method to the main component:
+
+```go
+type Main interface {
+    Main(context.Context) error
+}
+```
+
+And, we simplified `weaver.Run` to simply call this `Main` method. As a result,
+the signature of `weaver.Run` was much simpler:
+
+```go
+func Run(context.Context)
+```
+
+Here's the todo app again:
+
+```go
+func main() {
+    flag.Parse()
+    if err := weaver.Run(context.Background()); err != nil {
+        log.Fatal(err)
+    }
+}
+
+type app struct {
+    weaver.Implements[weaver.Main]
+    todo weaver.Ref[Todo]
+}
+
+func (a *app) Main(ctx context.Context) {
+    // ...
+}
+```
+
+## June 21, 2023: [Remove Main Method](https://github.com/ServiceWeaver/weaver/commit/a4a3f37f6ef443b853083e9627f7ec15ce5e3c9a)
+
+Service Weaver ships with a `weavertest` package that you can use to test a
+Service Weaver application. Here's how to test the `Todo` component, for
+example:
+
+```go
+func TestTodo(t *testing.T) {
+    weavertest.Local.Test(t, func(t *testing.T, todo Todo) {
+        // Test Todo here...
+    })
+}
+```
+
+`weavertest.Local.Test` is similar to `weaver.Run`. It runs a local copy of the
+Service Weaver application, initializing all components as necessary. It then
+calls the provided function with a handle to the `Todo` component.
+
+We felt it was unclear whether the `Test` function should run the `Main` method
+of the main component or not. Because it mirrors `weaver.Run`, there's an
+argument that it should. However, many tests don't want to or can't run the
+`Main` method. For example, if a `Main` method listens on a hardcoded address
+(e.g., `localhost:9000`), then two calls to `Test` cannot run in parallel if
+they both call the `Main` method. Ultimately, we decided to revert back to the
+more explicit `weaver.Run`:
+
+```go
+func main() {
+    flag.Parse()
+    if err := weaver.Run(context.Background(), serve); err != nil {
+        log.Fatal(err)
+    }
+}
+
+type app struct {
+    weaver.Implements[weaver.Main]
+    todo Todo
+}
+
+func (a *app) Init(context.Context) error {
+    a.todo = weaver.Get[Todo]()
+}
+
+func serve(ctx context.Context, app *app) {
+    // ...
+}
+```
+
+This version of `weaver.Run` clearly runs the provided function (e.g., `serve`),
+and `weavertest.Local.Test` clearly does not.
+
+## Lessons Learned
+
+This brings us to the Service Weaver API we have today.
```

Comment:
Can you add in some ideas in rough form (they can be fleshed out in a subsequent change):

* Using an API shows potential problems, so use it before committing to it
* API design is hard: full of pros and cons that need to be balanced. In particular, fine line between boilerplate and asking users to be precise.
* New mini-languages (//weaver:... comments) are simple to start with, but easy to get into something complicated
* Declaring things ahead of time can be powerful (e.g., the component graph)

Reply:
Added the "Lessons Learned" section you wrote.